docs/src/app/(private)/experiments/custom-components.tsx

```diff
@@ -0,0 +1,87 @@
+'use client';
+import * as React from 'react';
+import { useRenderer, RenderProp } from '@base-ui-components/react/use-renderer';
+import styles from './custom-components.module.css';
+
+type Weight = 'light' | 'regular' | 'bold';
+type Size = 'small' | 'medium' | 'large';
+
+type TextState = {
+  weight: Weight;
+  size: Size;
+  excludedProp: boolean;
+};
+
+type TextProps = {
+  className: string | ((state: TextState) => string);
+  weight?: Weight;
+  render?: RenderProp<TextState>;
+  children: React.ReactNode;
+  style?: React.CSSProperties;
+  size?: Size;
+  excludedProp?: boolean;
+};
+
+const Text = React.forwardRef(
+  (props: TextProps, forwardedRef: React.ForwardedRef<HTMLElement>) => {
+    const {
+      className,
+      render,
+      style = {},
+      weight = 'regular',
+      size = 'medium',
+      // Example state prop that we exclude from the style hooks
+      excludedProp = true,
+      ...otherProps
+    } = props;
+
+    const fontWeight = {
+      light: 300,
+      regular: 400,
+      bold: 700,
+    }[weight];
+
+    const state = React.useMemo(
+      () => ({ weight, size, excludedProp }),
+      [weight, size, excludedProp],
+    );
+
+    const { renderElement } = useRenderer({
+      render: render ?? 'p',
+      state,
+      className,
+      ref: forwardedRef,
+      props: {
+        ...otherProps,
+        style: {
+          ...style,
+          fontWeight,
+        },
+      },
```

Comment:
`ref` is considered a "regular" prop now. We do still need to support React 17-18, so not sure if sticking in the same object causes issues. Also, for `extraProps`, I was referring to spreading them inside the same top-level object passed to `useRenderer`

Reply:
I liked the separation, just so it's clear what's top level API vs other props spread, e.g. state is not passed on the element, but everything else is, it may be confusing.

Reply:
For ref, maybe yes, we can include it in the props, I don't see how it could cause issues even for React 17-18, we are still receiving it in the hook, it doesn't matter where it came from, no? @michaldudak can you think of any issue with this?

Reply:
> I liked the separation, just so it's clear what's top level API vs other props spread, e.g. state is not passed on the element, but everything else is, it may be confusing.

I like it as well. Plus, it removes the need to spread and extract props from the hook's parameters, so it should save a few CPU cycles.

As for the ref, I can't think of a scenario where it won't work when placed among other props.

Reply:
Move the ref to the props object.